scripts/triggers-end-to-end-tests/cli.ts

```diff
@@ -0,0 +1,69 @@
+import * as subprocess from "child_process";
+
+export class CLIProcess {
+  process?: subprocess.ChildProcess;
+
+  constructor(private readonly name: string, private readonly workdir: string) {}
+
+  start(
+    cmd: string,
+    project: string,
+    additionalArgs: string[],
+    logDoneFn?: (d: unknown) => unknown
+  ): Promise<void> {
+    const args = [cmd, "--project", project];
+
+    if (additionalArgs) {
+      args.push(...additionalArgs);
+    }
+
+    this.process = subprocess.spawn("firebase", args, { cwd: this.workdir });
+    if (!this.process) {
+      throw new Error("Failed to start firebase CLI");
+    }
+
+    this.process.stdout.on("data", (data: unknown) => {
+      process.stdout.write(`[${this.name} stdout] ` + data);
+    });
+
+    this.process.stderr.on("data", (data: unknown) => {
+      console.log(`[${this.name} stderr] ` + data);
+    });
+
+    let started: Promise<void>;
+    if (logDoneFn) {
+      started = new Promise((resolve) => {
+        this.process?.stdout.on("data", (data: unknown) => {
+          if (logDoneFn(data)) {
+            resolve();
+          }
+        });
+      });
+    } else {
+      started = new Promise((resolve) => {
+        this.process?.once("close", () => {
+          this.process = undefined;
+          resolve();
+        });
+      });
+    }
+
+    return started;
+  }
+
+  stop(): Promise<void> {
+    if (!this.process) {
+      return Promise.resolve();
+    }
+
+    const stopped = new Promise((resolve) => {
+      this.process?.once("exit", (/* exitCode, signal */) => {
```

Comment:
Why do we need the null safe stuff here if we checked `!this.process` above?

Reply:
The typescript compiler was giving me grief. I think it's right though, since => keeps the right scope for `this` (and promise functions have no guarantee for when they will be run), `this.process` _could_ be undefined here (though I find it unlikely).

I might be able to solidify it a bit by putting it into a new variable, but this works for the time being.

Reply:
I made it better hah.